// xpring-js/build/grpc-ilp-network-client.web.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const get_balance_request_pb_1 = require("./generated/web/ilp/get_balance_request_pb");
const send_payment_request_pb_1 = require("./generated/web/ilp/send_payment_request_pb");
const utils_1 = __importDefault(require("./utils"));
const balance_service_grpc_web_pb_1 = require("./generated/web/ilp/balance_service_grpc_web_pb");
const ilp_over_http_service_grpc_web_pb_1 = require("./generated/web/ilp/ilp_over_http_service_grpc_web_pb");
class GrpcIlpNetworkClientWeb {
    constructor(grpcURL) {
        if (utils_1.default()) {
            try {
                // This polyfill hack enables XMLHttpRequest on the global node.js state
                // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
                // @ts-ignore eslint-disable-line
                global.XMLHttpRequest = require('xhr2'); // eslint-disable-line
            }
            catch (_a) {
                // Swallow the error here for browsers
            }
        }
        // FIXME wrong credentials
        this.balanceClient = new balance_service_grpc_web_pb_1.BalanceServiceClient(grpcURL, null, null);
        this.paymentClient = new ilp_over_http_service_grpc_web_pb_1.IlpOverHttpServiceClient(grpcURL, null, null);
    }
    getBalance(request, bearerToken) {
        return new Promise((resolve, reject) => {
            const metaData = { Authorization: bearerToken || '' };
            this.balanceClient.getBalance(request, metaData, (error, response) => {
                if (error != null || response === null) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    send(request, bearerToken) {
        return new Promise((resolve, reject) => {
            const metaData = { Authorization: bearerToken || '' };
            this.paymentClient.sendMoney(request, metaData, (error, response) => {
                if (error != null || response === null) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    /* eslint-disable class-methods-use-this */
    SendPaymentRequest() {
        return new send_payment_request_pb_1.SendPaymentRequest();
    }
    GetBalanceRequest() {
        return new get_balance_request_pb_1.GetBalanceRequest();
    }
}
exports.default = GrpcIlpNetworkClientWeb;
