// xpring-js/build/grpc-network-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const grpc_1 = require("grpc");
const get_account_info_pb_1 = require("./generated/node/org/xrpl/rpc/v1/get_account_info_pb");
const get_fee_pb_1 = require("./generated/node/org/xrpl/rpc/v1/get_fee_pb");
const submit_pb_1 = require("./generated/node/org/xrpl/rpc/v1/submit_pb");
const xrp_ledger_grpc_pb_1 = require("./generated/node/org/xrpl/rpc/v1/xrp_ledger_grpc_pb");
const account_pb_1 = require("./generated/node/org/xrpl/rpc/v1/account_pb");
const utils_1 = __importDefault(require("./utils"));
const get_account_transaction_history_pb_1 = require("./generated/node/org/xrpl/rpc/v1/get_account_transaction_history_pb");
const get_transaction_pb_1 = require("./generated/node/org/xrpl/rpc/v1/get_transaction_pb");
/**
 * A GRPC Based network client.
 */
class GRPCNetworkClient {
    constructor(grpcURL) {
        if (utils_1.default()) {
            this.grpcClient = new xrp_ledger_grpc_pb_1.XRPLedgerAPIServiceClient(grpcURL, grpc_1.credentials.createInsecure());
        }
        else {
            throw new Error('Use gRPC-Web Network Client on the browser!');
        }
    }
    async getAccountInfo(request) {
        return new Promise((resolve, reject) => {
            this.grpcClient.getAccountInfo(request, (error, response) => {
                if (error != null || response == null) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    async getFee(request) {
        return new Promise((resolve, reject) => {
            this.grpcClient.getFee(request, (error, response) => {
                if (error != null || response == null) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    async getTransaction(request) {
        return new Promise((resolve, reject) => {
            this.grpcClient.getTransaction(request, (error, response) => {
                if (error != null || response == null) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    async submitTransaction(request) {
        return new Promise((resolve, reject) => {
            this.grpcClient.submitTransaction(request, (error, response) => {
                if (error != null || response == null) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    async getTransactionHistory(request) {
        return new Promise((resolve, reject) => {
            this.grpcClient.getAccountTransactionHistory(request, (error, response) => {
                if (error != null || response == null) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    /* eslint-disable class-methods-use-this */
    AccountAddress() {
        return new account_pb_1.AccountAddress();
    }
    GetAccountInfoRequest() {
        return new get_account_info_pb_1.GetAccountInfoRequest();
    }
    GetTransactionRequest() {
        return new get_transaction_pb_1.GetTransactionRequest();
    }
    GetFeeRequest() {
        return new get_fee_pb_1.GetFeeRequest();
    }
    SubmitTransactionRequest() {
        return new submit_pb_1.SubmitTransactionRequest();
    }
    GetAccountTransactionHistoryRequest() {
        return new get_account_transaction_history_pb_1.GetAccountTransactionHistoryRequest();
    }
}
exports.default = GRPCNetworkClient;
